python/tank/platform/environment.py

```diff
@@ -329,8 +328,24 @@ def get_app_descriptor(self, engine_name, app_name):
 
         return d
 
-    def set_context(self, context):
-        self.__context = context
+    def change_context(self, new_context):
+        """
+        Sets the environment's context and refreshes all environment data.
+
+        :param new_context: The new sgtk.context.Context to switch to.
+        """
+        self.__context = new_context
+
+        # We need to figure out the env file path before we can
+        # refresh. We run it through the new context through the
+        # core hook, which will give us the name (like shot_step),
+        # and then get the filesystem path to that .yml file.
+        sgtk = new_context.tank
+        env_name = sgtk.execute_core_hook(
+            constants.PICK_ENVIRONMENT_CORE_HOOK_NAME,
+            context=new_context,
+        )
+        self._env_path = sgtk.pipeline_configuration.get_environment_path(env_name)
```

Comment:
Right, I have been thinking about this, looking at the code a bit more. Here's my suggestion how we should do this:

It all begins with `engine.change_context()` calling `_get_env_and_descriptor_for_engine`. This calls `pick_environment()` so we know at this point what environment is appropriate for the new context that the engine is trying to switch to.

The engine now supports switching on the fly. The engine keeps an environment instance to track apps etc. This instance doesn't currently change. Instead, we call `change_context()` on the env object.

I think we should instead think of the environment object as immutable. (the WritableEnvironment object modifies itself, but it never changes the env file that it is pointing at). `pipelineconfig.get_environment(self, env_name, context=None, writable=False)` is the factory method for environments and the object we get back cannot change the environment_name nor the context. These are fixed. The context is merely an object we pass along so that the include files can resolve correctly as the env is loading. We drop the `change_context()/set_context()` method on the environment altogether. 

Instead, in `engine.change_context()`, we have the engine update its `self.__env` member with whatever is returned from `_get_env_and_descriptor_for_engine`. So the env object remains constant and the dynamic reshuffling all happens at the engine level.

Do you think that's a good approach? Again, apologies if I am missing something obvious! :)